### PR TITLE
Row-level TTL PR 1: new API

### PR DESCRIPTION
**Pull request by ableegoldman**
First PR for row-level ttl. Includes only the API changes and supporting code

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java [hunk 1]
```diff
@@ -56,7 +56,9 @@ public final class ResponsiveStores {
    * @return a supplier for a key-value store with the given options
    *         that uses Responsive's storage for its backend
    */
-  public static KeyValueBytesStoreSupplier keyValueStore(final ResponsiveKeyValueParams params) {
+  public static ResponsiveKeyValueBytesStoreSupplier keyValueStore(
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java:59]: @agavra FYI all the changes in ResponsiveStores are technically unrelated -- just holdovers from an alternative API approach I abandoned -- but I think it's the right thing to do anyways so I kept it in. Hope you don't mind

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+      }
+      return new TtlDuration(ttl, Ttl.FINITE);
+    }
+
+    // No ttl will be applied, in other words infinite retention
+    public static TtlDuration noTtl() {
+      return new TtlDuration(Duration.ZERO, Ttl.INFINITE);
+    }
+
+    // TODO(sophie): store ttl as long to avoid Duration conversions on the hot path
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java:162]: @agavra I'm actually a bit concerned about this. Even if we store it as millis we'll still have a Duration on the hot path since this is the return type. What if we just bite the bullet and have the user return a TTL in terms of millis rather than a Duration?

I know it's not as nice API-wise but To be honest I think most Streams users are probably pretty used to reasoning about time in terms of milliseconds. So I'm personally leaning towards replacing Duration with long millis entirely. Thoughts?

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java [hunk 3]
```diff
@@ -46,7 +46,16 @@ public static ResponsiveKeyValueParams fact(final String name) {
   }
 
   public ResponsiveKeyValueParams withTimeToLive(final Duration timeToLive) {
-    this.timeToLive = timeToLive;
+    return withTtlProvider(TtlProvider.withDefault(timeToLive));
+  }
+
+  public ResponsiveKeyValueParams withTtlProvider(final TtlProvider<?, ?> ttlProvider) {
+    if (ttlProvider.hasConstantTtl() && !ttlProvider.defaultTtl().isFinite()) {
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java:53]:
```suggestion
    if (ttlProvider.hasDefaultOnly() && !ttlProvider.defaultTtl().isFinite()) {
```
nit: `hasConstantTtl` trips me up a little bit, might as well just directly say what it is

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java [hunk 4]
```diff
@@ -46,7 +46,16 @@ public static ResponsiveKeyValueParams fact(final String name) {
...
+      throw new IllegalArgumentException("Passed in ttlProvider with infinite ttl, "
+                                             + "this is equivalent to no ttl and you "
+                                             + "should not set the ttlProvider for this");
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java:56]: just wondering, why not just do nothing (i.e. leave it as Optional.empty and just return early)? I'd rather not throw a bunch of errors as that makes it stricter in the way we allow usage (I actually like having "NOOP" instances instead of nulls/empty in Java, so as a user, especially in tests, I may consider doing that)

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+  /**
+   * Creates a new TtlProvider that has no default (equivalent to infinite retention for
+   * all records unless an override is specified).  Must be used in combination with
+   * exactly one of the {@link #fromKey(Function, Serde)}, {@link #fromValue(Function, Serde)},
+   * and {@link #fromKeyAndValue(BiFunction, Serde, Serde)} methods to define the row-level
+   * override function.
+   * <p>
+   * If no ttl whatsoever is desired for this store, simply don't configure a TtlProvider
+   * to begin with.
```
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java:55]: same comment as above, just leaving this here as a reminder to change the docs if you agree

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java (hunk 2 above)
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java:162]: Do we have evidence that this is actually significant on a per-retrieved doc basis? 

A few bitwise CPU operations seems to be dwarfed by the potential network call + the re-serialization in my opinion. Even if hits the in-memory cache, we're talking about loads from main memory into L1 cache.

The cost of creating a `Duration` is ~15ns (thanks ChatGPT for the estimation, looks legit to me, and half of that cost is just object allocation) and the cost of calling `toMillis` is ~1ns (assuming there's no nanosecond component), so we're talking 16ns per-key returned overhead for using Duration on a 3.2Ghz CPU.

**Review by agavra — APPROVED** [review]
LGTM! Just nits inline

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java (hunk 3 above)
Comment (ableegoldman) [reply kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java:53]: I guess to me, "has constant" is more "what it is" than "default only" hahaha. But I can also see it your way. I'll change it

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java (hunk 4 above)
Comment (ableegoldman) [reply kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java:56]: oh I actually would prefer to just silently convert it! I had the impression from previous interactions that you strongly preferred to throw an exception when the user does something that doesn't make sense rather than just correct it? Anyways I'm happy to change it

Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveKeyValueParams.java:56]: Now that you mention it... 😆 In cases where it's ambiguous I'd prefer throwing an error, but in this case I think we _are_ doing what they want, we're just optimizing it away underneath the hood. (i.e. it's not invalid to set a default + infinite TTL, it's just the same as the default behavior)
